pkg/processor/metricfrequencyprocessor/sieve.go

```diff
@@ -0,0 +1,197 @@
+	return func(dataPoint pdata.DoubleDataPoint) bool {
+		cachedPoints := fs.metricCache.List(name)
+		fs.metricCache.Register(name, dataPoint)
+		lastReported, exists := fs.lastReported[name]
+		if !exists {
+			fs.lastReported[name] = dataPoint.Timestamp()
+			return false
+		}
+		earliest := earliestTimestamp(cachedPoints)
+		cachedPoints[dataPoint.Timestamp()] = dataPoint.Value()
+
+		if metricRequiresSamples(dataPoint, earliest) {
+			fs.lastReported[name] = dataPoint.Timestamp()
+			return false
+		}
+
+		if pastCategoryFrequency(dataPoint, lastReported, category1ReportFrequency) {
+			fs.lastReported[name] = dataPoint.Timestamp()
+			return false
+		}
+
+		if isConstant(dataPoint, cachedPoints) {
+			return true
+		}
+
+		if pastCategoryFrequency(dataPoint, lastReported, category2ReportFrequency) {
+			fs.lastReported[name] = dataPoint.Timestamp()
+			return false
+		}
+
+		if isLowInformation(cachedPoints) {
+			return true
+		}
+
+		if pastCategoryFrequency(dataPoint, lastReported, category3ReportFrequency) {
+			fs.lastReported[name] = dataPoint.Timestamp()
+			return false
+		}
+
+		return true
+	}
+}
+
+func metricRequiresSamples(point pdata.DoubleDataPoint, earliest pdata.Timestamp) bool {
+	return point.Timestamp().AsTime().Before(earliest.AsTime().Add(minPointAccumulationTime))
+}
+
+func pastCategoryFrequency(point pdata.DoubleDataPoint, lastReport pdata.Timestamp, categoryFrequency time.Duration) bool {
+	return point.Timestamp().AsTime().After(lastReport.AsTime().Add(categoryFrequency).Add(safetyInterval))
+}
+
+func isConstant(point pdata.DoubleDataPoint, points map[pdata.Timestamp]float64) bool {
+	for _, value := range points {
+		if !almostEqual(point.Value(), value) {
+			return false
+		}
+	}
+
+	return true
+}
+
+// heuristic attempt at defining uninteresting metrics. Requirements:
+// 1) no big changes - defined by no iqr anomalies
+// 2) little oscillations - defined by low variation
+func isLowInformation(points map[pdata.Timestamp]float64) bool {
+	q1, q3 := calculateQ1Q3(points)
+	iqr := q3 - q1
+	variation := calculateVariation(points)
+
+	return withinBounds(points, q1-iqrAnomalyCoef*iqr, q3+iqrAnomalyCoef*iqr) && lowVariation(variation, iqr)
+}
+
+// refers to quantiles - .25 and .75 respectively
+func calculateQ1Q3(points map[pdata.Timestamp]float64) (float64, float64) {
+	values := valueSlice(points)
+	sort.Float64s(values)
+	q1Index := len(points) / 4
+	q3Index := 3 * len(points) / 4
+	return values[q1Index], values[q3Index]
+}
+
+func withinBounds(points map[pdata.Timestamp]float64, lowerBound float64, upperBound float64) bool {
+	for _, v := range points {
+		if v < lowerBound {
+			return false
+		}
+		if v > upperBound {
+			return false
+		}
+	}
+
+	return true
+}
+
+// sum of absolute values of differences of subsequent data points
+func calculateVariation(points map[pdata.Timestamp]float64) float64 {
+	keys := keySlice(points)
+	sortTimestampArray(keys)
+
+	variation := 0.0
+	previous := keys[0]
```

Comment:
What if `points` has 0 elements?

Reply:
Currently processed data point is added to the array, not to mention it will be filtered before reaching this code.